# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/quota/InitializationUtils.h
# Commit: 013f6b408358
# Full Hash: 013f6b4083587bb523626c2cff60b74ac4d05871
# Author: Jan Varga <jvarga@mozilla.com>
# Date: 2024-10-15 04:16:11
# Regressor Bug: 1866402
# File Overlap Count: 1
# Description:
#   Bug 1866402 - Further reduce code duplication in QuotaManager::OpenStorageDirectory and QuotaManager::OpenClientDirectory; r=dom-storage-reviewers,jari
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D223710
# ==============================================================================

diff -r 84592335a026 -r 013f6b408358 dom/quota/InitializationUtils.h
--- a/dom/quota/InitializationUtils.h	Mon Oct 14 10:06:14 2024 +0000
+++ b/dom/quota/InitializationUtils.h	Mon Oct 14 10:06:14 2024 +0000
@@ -14,9 +14,32 @@
 #include "mozilla/dom/quota/DirectoryLock.h"
 #include "mozilla/dom/quota/DirectoryLockInlines.h"
 #include "mozilla/dom/quota/ForwardDecls.h"
+#include "mozilla/dom/quota/QuotaManager.h"
 
 namespace mozilla::dom::quota {
 
+template <typename UninitChecker, typename PromiseArrayIter>
+RefPtr<UniversalDirectoryLock> CreateDirectoryLockForInitialization(
+    QuotaManager& aQuotaManager, const PersistenceScope& aPersistenceScope,
+    const OriginScope& aOriginScope, const bool aAlreadyInitialized,
+    UninitChecker&& aUninitChecker, PromiseArrayIter&& aPromiseArrayIter) {
+  RefPtr<UniversalDirectoryLock> directoryLock =
+      aQuotaManager.CreateDirectoryLockInternal(aPersistenceScope, aOriginScope,
+                                                Nullable<Client::Type>(),
+                                                /* aExclusive */ false);
+
+  if (aAlreadyInitialized &&
+      !std::forward<UninitChecker>(aUninitChecker)(directoryLock)) {
+    return nullptr;
+  }
+
+  auto iter = std::forward<PromiseArrayIter>(aPromiseArrayIter);
+  *iter = directoryLock->Acquire();
+  ++iter;
+
+  return directoryLock;
+}
+
 template <typename Callable>
 class MaybeInitializeHelper final {
  public:
